### DISCUSSION TITLE
Ideas: Phase 7 directions — MAML/Reptile meta-learning, episodic replay, or hypernetworks?

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#258] [Ideas]
## Setting the stage for Phase 7

Phase 6 is wrapping up the EWC continual-learning system (sub-phases 6.1–6.5). It's a good moment to sketch Phase 7 options and gather input before committing to a direction.

---

## What Phase 6 gives us

By the end of Phase 6.5, the agent will have:

- **Persistent parameter importance** — Fisher matrices per task, stored in Neo4j/Redis
- **SLEEP_PHASE consolidation** — snapshot capture + EWC regularisation during rest cycles
- **Online Fisher accumulation** — EMA-based incremental updates (no full-batch recomputation)
- **Multi-task routing** — `TaskRegistry` + `TaskContextManager` for context-aware EWC

This gives us a solid **continual learning** foundation. The agent won't forget as it learns new tasks.

---

### Option A: MAML-style meta-learning (Model-Agnostic Meta-Learning)

**What**: Train the agent to learn new tasks quickly from few examples by optimising for fast adaptation. The agent learns a prior that is "easy to fine-tune."

**Integration point**: `STDPOnlineLearner` gains a `meta_update()` method that computes second-order gradient updates across a task distribution.

**Pros**:
- Directly addresses few-shot generalisation (a core ASI goal)
- Compatible with EWC (meta-prior + importance weights are complementary)
- Well-studied: Finn et al. 2017, Nichol et al. 2018 (Reptile, first-order approx)

**Cons**:
- Requires second-order gradients or a first-order approximation (Reptile is simpler)
- Task distribution must be defined — harder in open-ended agent settings

---

### Option B: Episodic replay with priority sampling

**What**: Store a replay buffer of past episodes (in the existing KG episodic memory) and interleave past experiences into the LEARNING phase to prevent forgetting — without needing explicit Fisher matrices.

**Integration point**: `MemoryConsolidator` gains a `sample_replay()` method; `CognitiveCycle` adds a `REPLAY` phase between `LEARNING` and `INTEGRATION`.

**Pros**:
- Conceptually simpler than meta-learning
- Leverages existing KG episodic storage (Phase 5)
- Empirically strong baseline (Prioritized Experience Replay — PER, Schaul et al. 2015)

**Cons**:
- Memory-hungry: replay buffer can grow unboundedly
- Privacy implications: raw episode replay may leak sensitive past inputs

---

### Option C: Hypernetwork-based task conditioning

**What**: A small "hypernetwork" generates task-specific weight offsets given a task embedding, rather than modifying the base parameter vector directly.

**Integration point**: New `HyperController` module generates `delta_W` for each task; `STDPOnlineLearner` applies `W_effective = W_base + delta_W(task_embed)`.

**Pros**:
- Cleanest architectural separation between shared knowledge and task-specific adaptation
- No Fisher matrices needed (task-specific knowledge is explicit in hypernetwork weights)
- Studied in Ha & Schmidhuber 2016, von Oswald et al. 2019

**Cons**:
- Adds significant architectural complexity
- Requires a task embedding mechanism (how do you represent "task_id=xyz" as a vector?)

---

### Option D: Consolidate + harden Phase 6 before new features

Before adding Phase 7, fully implement Phase 6 in code:

- Implement all 5 Phase 6 sub-phases (6.1–6.5) as actual Python files
- Run the 50+ test targets
- Benchmark Fisher update latency end-to-end
- Write integration guide for external contributors

**Pros**:
- The specs are deep; implementation may surface design changes
- External contributors can join more easily with concrete code

**Cons**:
- Delays meta-learning work; doesn't generate new engagement on its own

---

## Preliminary recommendation

**Option A (Reptile first-order MAML)** is the strongest Phase 7 candidate:

1. Reptile is simple to implement: `theta_new = theta + eps * (theta_task - theta_base)` — no second-order gradients
2. It composes with EWC: the Fisher regulariser runs inside the inner-loop adaptation
3. It directly targets the "few-shot generalisation" capability that distinguishes ASI from specialist models

A minimal Phase 7 would be:
- `ReptileLearner` — outer-loop update rule
- `TaskSampler` — samples task batches from `TaskRegistry`
- Integration with SLEEP_PHASE (outer-loop update happens at rest boundary)

---

## Open questions

1. Does the community prefer **Option A (MAML/Reptile)** or **Option B (episodic replay)** for Phase 7? Both are compatible with Phase 6.
2. Should Phase 7 require implementing Phase 6 in actual code first, or can it proceed as spec-only (as Phases 5 and 6 have been)?
3. What task distribution should the `TaskSampler` use — drawn from the KG episode store, from a synthetic benchmark (Omniglot/MiniImageNet), or from the live environment?

Please share your thoughts below!

**Comment by web3guru888** [comment]
## Phase 7.1 Filed — Reptile Meta-Learning Foundation

Following the discussion here, **Reptile** has been selected as the Phase 7.1 starting point (Option A first-order, not MAML):

- **Issue #259** — Phase 7.1: `ReptileMetaLearner`, `TaskSample`, inner-loop adaptation, 14 test targets
- **Discussion #260** — Show & Tell: Reptile algorithm, component map, SLEEP_PHASE integration, EWC+Reptile interaction
- **Discussion #261** — Q&A: `inner_steps`, `meta_lr`, `task_batch_size`, snapshot/restore, convergence timeline
- **Wiki** — Phase-7-Meta-Learning.md (see below for wiki)

### Rationale recap

Reptile was chosen over MAML because:
- No second-order gradients (no Hessian) → works in pure numpy, no autograd
- Runs within SLEEP_PHASE budget without blowing up compute
- Naturally compatible with `FisherAccumulator` (EWC penalty in inner loop)
- `TaskRegistry` from Phase 6.5 provides per-task context routing with zero changes

### Updated Phase 7 sub-phase tracker

| Sub-phase | Component | Status |
|---|---|---|
| **7.1** | `ReptileMetaLearner`, `TaskSample`, `MetaParamSnapshot` | 🔵 Open (#259) |
| 7.2 | Task distribution sampler — `TaskSampler`, curriculum scheduling | Planned |
| 7.3 | Few-shot evaluation harness — N-way K-shot accuracy metrics | Planned |
| 7.4 | Meta-checkpoint store — persist/restore meta-params across restarts | Planned |
| 7.5 | Meta-learning integration tests — full SLEEP_PHASE cycle with Reptile | Planned |

### SLEEP_PHASE hook order (cumulative)

```
1. FisherAccumulator.force_snapshot()   [Phase 6.4]
2. EWCRegulariser.consolidate()         [Phase 6.1]
3. TaskRegistry.evict_lru()            [Phase 6.5]
4. ReptileMetaLearner.meta_update()    [Phase 7.1]  ← now specified
```
